### ISSUE TITLE
Compiler Hardening Guide should include -fstrict-flex-arrays

### ISSUE THREAD

**Issue by thomasnyman** [ossf/wg-best-practices-os-developers#249]
Brought up during the [C/C++ Compiler BP Call 2023-09.27](https://docs.google.com/document/d/1UClGUnOSkOH_wab6Lx43KUdkaK37L8sbWJ_GPZvc1YY/edit#bookmark=id.863n4hp4nw9q).

By default, GCC and Clang treat all trailing arrays (arrays that are placed as the last member or a structure) as flexible-sized arrays regardless of their actual size for the purposes of `__builtin_object_size()` calculations used by `__FORTIFY_SOURCE`:

```C
struct flex_array {
    int a;
    int b;
    int c[];
};

struct trailing_array {
    int a;
    int b;
    int c[4];
};

struct flex_array *flex;
struct trailing_array *trailing;

/* ... */

__builtin_object_size(flex, 1) == -1
__builtin_object_size(flex->c, 1) == -1
__builtin_object_size(trailing, 1) == -1
__builtin_object_size(trailing->c, 1) == -1  // trailing->c statically sized to 16 bytes
```
(Complete example modified from [original by Kees Cook](https://gcc.gnu.org/bugzilla/show_bug.cgi?id=101836) at https://godbolt.org/z/M93Yx53P4)

The rationale for the default behavior is to allow for the ['_struct hack_' idiom](https://www.geeksforgeeks.org/struct-hack/) that allows for trailing arrays to be treated as variable sized (regardless of their declared size):

```C
struct fam {  /* flexible array member */
    size_t name_len;
    char name[1];
}

/* ... */

struct fam* f = malloc(sizeof(struct fam) + sizeof(char[n]) - sizeof(char[1]));

/* equivalent to:
struct fam {
    size_t name_len;
    char name[n];
}
*/
```

The `-fstrict-flex-arrays=[0|1|2|3]` option makes the compiler respect the sizes of trailing array member more strictly. This allows bounds checks added by instrumentation such as `__FORTIFY_SOURCE` or `-fsanitize=bounds` to be able to correctly determine the size of trailing arrays. The tradeoff is that code that relies on the 'struct hack' for arbitrary sized trailing arrays may break as a result.

This flag is supported since [GCC 13](https://gcc.gnu.org/gcc-13/changes.html) and [Clang 15.0.0](https://releases.llvm.org/15.0.0/tools/clang/docs/ReleaseNotes.html#major-new-features).

## Considerations and references

- [GCC Bug 101835 - __builtin_object_size(P->M, 1) where M is an array and the last member of a struct fails](https://gcc.gnu.org/bugzilla/show_bug.cgi?id=101836) 
- [Patch and thread on GCC mailing list](https://gcc.gnu.org/pipermail/gcc-patches/2022-October/602902.html)
- [GCC documentation](https://gcc.gnu.org/onlinedocs/gcc-13.1.0/gcc/C-Dialect-Options.html#index-fstrict-flex-arrays)
- [Thread in LLVM Phabricator](https://reviews.llvm.org/D126864)
- [Serge Guelton's blog post on the benefits and limitations of flexible array members](https://developers.redhat.com/articles/2022/09/29/benefits-limitations-flexible-array-members#)

**Comment by david-a-wheeler** [comment]
FYI, here's an LWN article about flexible arrays: https://lwn.net/Articles/946041/

**Comment by david-a-wheeler** [comment]
I believe clang supports `-fstrict-flex-arrays` but only 0..2. At least in 2022 they decided to *not* implement level 3. I think they should have, but I don't make their decisions. :-). More info here: https://lwn.net/Articles/908817/

However, that's last years' information. Have things changed since?

**Comment by eslerm** [comment]
@david-a-wheeler I checked the commit in https://lwn.net/Articles/921871/ to add level 3 in llvm-project, and it's included from git tag llvmorg-16.0.0 onward

**Comment by kees** [comment]
> I believe clang supports `-fstrict-flex-arrays` but only 0..2. At least in 2022 they decided to _not_ implement level 3. I think they should have, but I don't make their decisions. :-). More info here: https://lwn.net/Articles/908817/
> 
> However, that's last years' information. Have things changed since?

Yes, `=3` is supported by Clang. It was only a matter of days/weeks where it didn't. :)

**Comment by david-a-wheeler** [comment]
I think we should recommend level 3, and state that it's available in clang 16.0.0 and up. In the notes we can note that levels 0..2 are in clang 15. I'll create a PR to propose it.

**Comment by david-a-wheeler** [comment]
I've created a PR that I believe resolves this: https://github.com/ossf/wg-best-practices-os-developers/pull/263
